test/e2e/network/traffic_distribution.go

```diff
@@ -0,0 +1,269 @@
+				e2epod.SetNodeSelection(&pod.Spec, nodeSelection)
+				pod.Labels = servingPodLabels
+
+				servingPods = append(servingPods, pod)
+				zoneForServingPod[pod.Name] = zone
+				ginkgo.DeferCleanup(framework.IgnoreNotFound(c.CoreV1().Pods(f.Namespace.Name).Delete), pod.GetName(), metav1.DeleteOptions{})
+			}
+			e2epod.NewPodClient(f).CreateBatch(ctx, servingPods)
+
+			trafficDist := v1.ServiceTrafficDistributionPreferClose
+			svc := createServiceReportErr(ctx, c, f.Namespace.Name, &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "traffic-dist-test-service",
+				},
+				Spec: v1.ServiceSpec{
+					Selector:            servingPodLabels,
+					TrafficDistribution: &trafficDist,
+					Ports: []v1.ServicePort{{
+						Port:       80,
+						TargetPort: intstr.FromInt32(9376),
+						Protocol:   v1.ProtocolTCP,
+					}},
+				},
+			})
+			ginkgo.By(fmt.Sprintf("creating a service=%q with trafficDistribution=%v", svc.GetName(), *svc.Spec.TrafficDistribution))
+			ginkgo.DeferCleanup(framework.IgnoreNotFound(c.CoreV1().Services(f.Namespace.Name).Delete), svc.GetName(), metav1.DeleteOptions{})
+
+			ginkgo.By("ensuring EndpointSlice for service have correct same-zone hints")
+			gomega.Eventually(ctx, endpointSlicesForService(svc.GetName())).WithPolling(5 * time.Second).WithTimeout(e2eservice.ServiceEndpointsTimeout).Should(endpointSlicesHaveSameZoneHints)
+
+			ginkgo.By("keeping traffic within the same zone as the client, when serving pods exist in the same zone")
+
+			createClientPod := func(ctx context.Context, zone string) *v1.Pod {
+				pod := e2epod.NewAgnhostPod(f.Namespace.Name, "client-pod-in-"+zone, nil, nil, nil)
+				pod.Spec.NodeName = nodeForZone[zone]
+				nodeSelection := e2epod.NodeSelection{Name: nodeForZone[zone]}
+				e2epod.SetNodeSelection(&pod.Spec, nodeSelection)
+				cmd := fmt.Sprintf(`date; for i in $(seq 1 3000); do sleep 1; echo "Date: $(date) Try: ${i}"; curl -q -s --connect-timeout 2 http://%s:80/ ; echo; done`, svc.Name)
+				pod.Spec.Containers[0].Command = []string{"/bin/sh", "-c", cmd}
+				pod.Spec.Containers[0].Name = pod.Name
+
+				ginkgo.DeferCleanup(framework.IgnoreNotFound(c.CoreV1().Pods(f.Namespace.Name).Delete), pod.GetName(), metav1.DeleteOptions{})
+				return e2epod.NewPodClient(f).CreateSync(ctx, pod)
+			}
+
+			for _, clientZone := range zones[:2] {
+				framework.Logf("creating a client pod for probing the service from zone=%q which also has a serving pod", clientZone)
+				clientPod := createClientPod(ctx, clientZone)
+
+				framework.Logf("ensuring that requests from clientPod=%q on zone=%q stay in the same zone", clientPod.Name, clientZone)
+
+				requestsSucceedAndStayInSameZone := gcustom.MakeMatcher(func(reverseChronologicalLogLines []string) (bool, error) {
+					logLines := reverseChronologicalLogLines
+					if len(logLines) < 20 {
+						return false, fmt.Errorf("got %d log lines, waiting for at least 10", len(logLines))
```

Comment:
```suggestion
						return false, fmt.Errorf("got %d log lines, waiting for at least 20", len(logLines))
```

shoulf you not try to retry here until you have enough log lines?

Reply:
I don't like this cryptic libraries :( , but I leave implementers to choose their own poison 😄

Reply:
Hahaha!

I totally get you :)